micropro: init at 0.4

kaliumxyz opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
getting the Micropro TL866CS universal programmer utility in nixpkgs

###### Things done
Wrote nix file for micropro utility, added self as its maintainer.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

kaliumxyz committed:
maintainers: add kalium

kaliumxyz committed:
micropro: init at 0.4

minipro is a utility for using the minipro TL866 series of universal programmers

corrected with suggestions from @IvarWithoutBones

IvarWithoutBones commented on pkgs/development/tools/minipro/default.nix:
```diff
@@ -0,0 +1,44 @@
+{stdenv, fetchgit, pkgs, pkg-config}:
+stdenv.mkDerivation rec {
+  name = "minipro-0.4";
```
Should be using `pname` and `version`.
```suggestion
  pname = "minipro";
  version = "0.4";
```

IvarWithoutBones commented on pkgs/development/tools/minipro/default.nix:
```diff
@@ -0,0 +1,44 @@
+    rev = "896fe7d9d6f1e77931626d273e880856884ee6d8";
+  };
+
+  PKG_CONFIG = "${pkg-config}/bin/pkg-config";
```
I don't see you using this anywhere, is this really required?

kaliumxyz replied:
Yes, the makeFile fails to find pkg-config, adding it here will set the environment variable and bypass it looking for the binary `pkg-config`

IvarWithoutBones commented on pkgs/development/tools/minipro/default.nix:
```diff
@@ -0,0 +1,44 @@
+stdenv.mkDerivation rec {
+  name = "minipro-0.4";
+
+  src = fetchGit {
```
```suggestion
  src = fetchgit {
    url = "https://gitlab.com/DavidGriffith/minipro.git";
    rev = "896fe7d9d6f1e77931626d273e880856884ee6d8";
    sha256 = "1jz85vzg5bb20hfyy5gbhj4v5n5iwc5rpcmpmx342q2g12zjqdsf";
  };
```

IvarWithoutBones commented on pkgs/development/tools/minipro/default.nix:
```diff
@@ -0,0 +1,44 @@
+
+  PKG_CONFIG = "${pkg-config}/bin/pkg-config";
+
+  nativeBuildInputs = with pkgs; [
```
You shouldn't use `with pkgs;` here, import them at the top instead. You should drop the `pkgs` attribute all together.

IvarWithoutBones commented on pkgs/development/tools/minipro/default.nix:
```diff
@@ -0,0 +1,44 @@
+    pkg-config
+  ];
+
+  buildInputs = with pkgs; [
```
ditto

IvarWithoutBones commented on pkgs/development/tools/minipro/default.nix:
```diff
@@ -0,0 +1,44 @@
+{stdenv, fetchgit, pkgs, pkg-config}:
```
```suggestion
{stdenv, fetchgit, pkg-config, libusb1 }:
```

IvarWithoutBones reviewed (commented):
Added a few suggestions :)

Also, commit message does not follow [the contributing guidelines](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

kaliumxyz commented:
@IvarWithoutBones I've incorporated your suggestions c:

IvarWithoutBones reviewed (commented):
Few more minor nits, otherwise LGTM :)

kaliumxyz commented:
@IvarWithoutBones I've split up the package in stable and unstable, moved to using leaveDotGit so that the package can be build as intended (which sadly does mean that we need to keep using fetchgit), fixed the manpages and the udev files, it should be good to go now c:

IvarWithoutBones commented:
> I've split up the package in stable and unstable

Why both? Wouldn't just stable be good enough?

> moved to using leaveDotGit so that the package can be build as intended

That's probably not a very good idea, because of #8567 (Also, pretty sure `fetchFromGitLab` has a similar flag). From local testing i noticed the testing isn't required, couldn't we just drop it? 

> Thank you for your help and feedback so far!

Glad it's helpful :) Apologies about inconsistencies and such, I'm quite a beginner at this 😅

Also, looks like you forgot to push.

kaliumxyz commented:
>Why both? Wouldn't just stable be good enough?
The unstable variant has support for newer firmware (though not for the minipro I use), and some fixes

>That's probably not a very good idea, because of #8567 (Also, pretty sure fetchFromGitLab has a similar flag). [...]
The `.git` is for version information, its not required but its part of the tool, I could not find a similar flag on `fetchFromGitLab`, seeing that its only used for version information and not part of the output, it should not lead to issues with the build being nondeterministic. I just realized that the build output includes a timestamp in the version, which should probably be patched out (let me do that quickly).

>forgot to push
pushed now c:

IvarWithoutBones commented:
> The unstable variant has support for newer firmware (though not for the minipro I use), and some fixes

Ah, okay :+1: 

> seeing that its only used for version information and not part of the output, it should not lead to issues with the build being nondeterministic

This is an issue in the cloning process, so it would.

IvarWithoutBones commented on pkgs/development/tools/minipro/unstable.nix:
```diff
@@ -0,0 +1,49 @@
+{ stdenv, installShellFiles, which, fetchgit, git, pkg-config, libusb1 }:
+stdenv.mkDerivation rec {
+  pname = "minipro";
+  version = "0.4dev";
```
Version naming convention for unstable git revs is `unstable-<date>`.
```suggestion
  version = "unstable-2020-04-18";
```

kaliumxyz commented:
we can go back to patching out the version information and just showing blanks on `-V`, setting it to anything else relevant, or manually setting it for every update (unless there is some other way we can get the information of the current commit?).

IvarWithoutBones commented on pkgs/development/tools/minipro/unstable.nix:
```diff
@@ -0,0 +1,49 @@
+    substituteInPlace Makefile --replace '$(shell date "+%Y-%m-%d %H:%M:%S %z")' "1970-01-01 00:00:00 +0000"
+  '';
+
+  buildPhase = ''
```
Could be using `makeFlags` for this.
```suggestion
  makeFlags = [ "-e minipro" ];
```

IvarWithoutBones commented on pkgs/development/tools/minipro/unstable.nix:
```diff
@@ -0,0 +1,49 @@
+  ];
+
+  buildInputs = [
+    stdenv
```
`stdenv` doesn't seem to be required

IvarWithoutBones commented on pkgs/development/tools/minipro/unstable.nix:
```diff
@@ -0,0 +1,49 @@
+    description = "An open source program for controlling the MiniPRO TL866xx series of chip programmers";
+    homepage = "https://gitlab.com/DavidGriffith/minipro";
+    maintainers = with maintainers; [ kalium ];
+    license = stdenv.lib.licenses.gpl3;
```
Since you're already doing `with stdenv.lib;`, you don't need to mention that here.
```suggestion 
    license = licenses.gpl3;
```

IvarWithoutBones commented on pkgs/development/tools/minipro/default.nix:
```diff
@@ -0,0 +1,49 @@
+{ stdenv, installShellFiles, which, fetchgit, git, pkg-config, libusb1 }:
```
It doesn't look like you're using either `git` or `which`, and the package seems to compile fine without them.

IvarWithoutBones reviewed (commented):
Few more nits :)

IvarWithoutBones commented:
> we can go back to patching out the version information and just showing blanks on -V

I have [a local expression](https://gist.github.com/IvarWithoutBones/f3b9c72c1fff7f05cb76407027d08875) which does none of the git patching, and doesn't have `leaveDotGit` enabled. However, it still outputs the following when executing it with the `-V` flag:
```
minipro version 0.4dev     A free and open TL866XX programmer
Build date:     2020-06-12 07:43:54 +0000
Commit date:    2019-10-19 23:38:37 +0000
Git commit:     5c397ebf4b5e736c31f4469821895bcc1bd38d69
Git branch:     tag: 0.4, refs/keep-around/5c397ebf4b5e736c31f4469821895bcc1bd38d69
```

I think we don't need to worry about that at all.

kaliumxyz commented:
okay good, seems to work for me as well now.

I've incorporated your comments, and only patch out the buildDate timestamp (else its not deterministic) :).

IvarWithoutBones reviewed (approved):
LGMT :)
